kv/kvserver: TestNonTxnReadWithinUncertaintyIntervalAfterLeaseTransfer failed

kv/kvserver.TestNonTxnReadWithinUncertaintyIntervalAfterLeaseTransfer [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/15586654?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/15586654?buildTab=artifacts#/) on release-23.2 @ [10a44bdde363db0623b7dc794167941d49f1c3c8](https://github.com/cockroachdb/cockroach/commits/10a44bdde363db0623b7dc794167941d49f1c3c8):


```
=== RUN   TestNonTxnReadWithinUncertaintyIntervalAfterLeaseTransfer
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestNonTxnReadWithinUncertaintyIntervalAfterLeaseTransfer1141057536
    test_log_scope.go:81: use -show-logs to present logs inline
    client_replica_test.go:1203: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_replica_test.go:1203
        	Error:      	Should be true
        	Test:       	TestNonTxnReadWithinUncertaintyIntervalAfterLeaseTransfer
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestNonTxnReadWithinUncertaintyIntervalAfterLeaseTransfer1141057536
--- FAIL: TestNonTxnReadWithinUncertaintyIntervalAfterLeaseTransfer (51.44s)
```

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestNonTxnReadWithinUncertaintyIntervalAfterLeaseTransfer.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-39451

--- a/pkg/kv/kvserver/client_replica_test.go
+++ b/pkg/kv/kvserver/client_replica_test.go
@@ -1143,9 +1143,7 @@ func TestNonTxnReadWithinUncertaintyIntervalAfterLeaseTransfer(t *testing.T) {
 	var maxNanos int64
 	for _, m := range manuals {
 		m.Pause()
-		if cur := m.UnixNano(); cur > maxNanos {
-			maxNanos = cur
-		}
+		maxNanos = max(maxNanos, m.UnixNano())
 	}
 	// After doing so, perfectly synchronize them.
 	for _, m := range manuals {
@@ -1180,8 +1178,12 @@ func TestNonTxnReadWithinUncertaintyIntervalAfterLeaseTransfer(t *testing.T) {
 		t.Fatalf("timeout")
 	}
 
-	// Advance the clock on node 1.
+	// Advance the clock on node 1. This should now lead the clock on node 2 and
+	// the timestamp assigned to the non-txn read, because the two manual clocks
+	// were paused and synchronized up above.
 	manuals[0].Increment(100)
+	clockTs := tc.Servers[0].Clock().Now()
+	require.True(t, nonTxnOrigTs.Less(clockTs), "nonTxnOrigTs: %v, clockTs: %v", nonTxnOrigTs, clockTs)
 
 	// Perform a non-txn write on node 1. This will grab a timestamp from node 1's
 	// clock, which leads the clock on node 2 and the timestamp assigned to the
@@ -1199,11 +1201,12 @@ func TestNonTxnReadWithinUncertaintyIntervalAfterLeaseTransfer(t *testing.T) {
 	// operations and assert that we observe an uncertainty error even though its
 	// absence would not be a true stale read.
 	ba := &kvpb.BatchRequest{}
+	ba.RangeID = desc.RangeID
 	ba.Add(putArgs(key, []byte("val")))
-	br, pErr := tc.Servers[0].DistSenderI().(kv.Sender).Send(ctx, ba)
+	br, pErr := tc.GetFirstStoreFromServer(t, 0).Send(ctx, ba)
 	require.Nil(t, pErr)
 	writeTs := br.Timestamp
-	require.True(t, nonTxnOrigTs.Less(writeTs))
+	require.True(t, nonTxnOrigTs.Less(writeTs), "nonTxnOrigTs: %v, writeTs: %v", nonTxnOrigTs, writeTs)
 
 	// Then transfer the lease to node 2. The new lease should end up with a start
 	// time above the timestamp assigned to the non-txn read.

